lib/spack/spack/environment/environment.py

```diff
@@ -1111,14 +1117,50 @@ def _concretize_separately(self, tests=False):
                 self._add_concrete_spec(s, concrete, new=False)
 
         # Concretize any new user specs that we haven't concretized yet
-        concretized_specs = []
+        arguments, root_specs = [], []
         for uspec, uspec_constraints in zip(
-                self.user_specs, self.user_specs.specs_as_constraints):
+                self.user_specs, self.user_specs.specs_as_constraints
+        ):
             if uspec not in old_concretized_user_specs:
-                concrete = _concretize_from_constraints(uspec_constraints, tests=tests)
-                self._add_concrete_spec(uspec, concrete)
-                concretized_specs.append((uspec, concrete))
-        return concretized_specs
+                root_specs.append(uspec)
+                arguments.append((uspec_constraints, tests))
+
+        # Ensure we don't try to bootstrap clingo in parallel
+        if spack.config.get('config:concretizer') == 'clingo':
+            with spack.bootstrap.ensure_bootstrap_configuration():
+                spack.bootstrap.ensure_clingo_importable_or_raise()
+
+        # Ensure all the indexes have been built or updated, since
+        # otherwise the processes in the pool may timeout on waiting
+        # for a write lock. We do this indirectly by retrieving the
+        # provider index, which should in turn trigger the update of
+        # all the indexes if there's any need for that.
+        _ = spack.repo.path.provider_index
+
+        # Ensure we have compilers in compilers.yaml to avoid that
+        # processes try to write the config file in parallel
+        _ = spack.compilers.get_compiler_config()
+
+        # Solve the environment in parallel on Linux
+        start = time.time()
+        if sys.platform != 'darwin':
+            nprocesses = min(spack.util.cpus.cpus_available(), 16)
```

Comment:
might as well cap it with the number of root specs too
```suggestion
            nprocesses = min(len(root_specs), min(spack.util.cpus.cpus_available(), 16))
```

in some cases we avoid creating a pool if `nprocesses == 1` and just do a normal function call. maybe that applies here too? stacktraces would look better than, even with your proposed fixes to these from this pr. in particular relevant if len(root_specs) == 1.

Reply:
Also debugging is quite a bit easier if we can avoid the pool...

Reply:
Not sure about that. I agree if we want to optimize and cap on `len(root_specs)` but having two ways to perform concretization (in a sub-process vs. directly in the parent) seems also to entail having two code paths to debug and test. `macOS` is still serial because there are issues with passing environments across processes that are difficult to nail down and this PR already provides a good speed-up for pipelines (so I think it's worth merging a linux only support at the moment).

Reply:
Capped on the number of specs that we need to concretize

Reply:
Here's an example from the small "build-systems" pipeline:
![Screenshot from 2021-10-12 14-23-30](https://user-images.githubusercontent.com/4199709/136955432-e56862e8-c1d7-4925-b6d2-680d84ac408d.png)